### DISCUSSION TITLE
Need help resolving issues with bar chart

### DISCUSSION THREAD

**Discussion by softwareDev-git** [recharts/recharts#4701] [Q&A] [answered]
I have to display a bar chart for timestamp (x-axis) and count (y-axis) data where x-axis has a scale of time.

I want to clarify if the following behaviours can be resolved -

1. **Show axis lines and grid lines for overflowing bar data.** For example, the below bar chart has bar data for June 16, June 17 and June 18. However, axis lines and grid lines are not drawn to accommodate the June 18 bar data. Is there a way to resolve axis line and grid lines here?

<img width="802" alt="image" src="https://github.com/recharts/recharts/assets/161946189/ee6eb861-71be-49a8-a98c-e04885f54060">

\
\
2. **Adjust tooltip position when x-axis has a scale of time but a type of category.** When using the time scale and number type for the x-axis, the tooltip behaves correctly when hovering over the bars. However, when using the time scale and category type for the x-axis, the tooltip doesn't behave correctly.

For example, the below bar chart is the same as the above except it has category type for its x-axis. The use of category type is needed to display the bars starting from one tick and ending at another. However, the tooltip for June 17 bar is displayed when the cursor hovers over the June 16 bar where the white circle is drawn. Is there a way to resolve tooltip position here?

<img width="794" alt="image" src="https://github.com/recharts/recharts/assets/161946189/005c6e0c-22a3-4af8-88a6-3cde89d4018f">

\
Please see codesandbox which displays these examples - [Codesandbox for Recharts](https://codesandbox.io/p/sandbox/sandbox-for-recharts-chqn82?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxuc7nfe0006356ixtqlaplj%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxuc7nfe0002356ifef9yoys%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxuc7nfe0003356izcrp2hpp%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxuc7nfe0005356i8dwz04b4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxuc7nfe0002356ifef9yoys%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxuc7nfe0001356i6rpce7sf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxuc7nfe0002356ifef9yoys%2522%252C%2522activeTabId%2522%253A%2522clxuc7nfe0001356i6rpce7sf%2522%257D%252C%2522clxuc7nfe0005356i8dwz04b4%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxuc7nfe0004356ipr4aqib6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxuc7nfe0005356i8dwz04b4%2522%252C%2522activeTabId%2522%253A%2522clxuc7nfe0004356ipr4aqib6%2522%257D%252C%2522clxuc7nfe0003356izcrp2hpp%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxuc7nfe0003356izcrp2hpp%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

\
I am relatively new to Recharts so any guidance or suggestion is appreciated. Thanks.

**Comment by ckifer** [comment, marked as answer]
Hi! Thanks for making the codesandbox!

`scale="time"` and `type="category"` are incompatible. d3 scale expects numerical data when using a timescale, or else data cannot be interpreted as a date. If you remove `scale="time"` the categorical chart works fine and is just that, a list of categories. As long as your data is in order and you don't need to interpolate this should work for your case. This is generally preferred with bar charts.

**Reply by softwareDev-git** [reply]
Thanks @ckifer for the explanation.

Can you help with point 1? I am using x-axis with scale of time and type of number but bar data still overflows.

**Reply by ckifer** [reply]
Generally I wouldn't recommend bar charts with a numeric XAxis, but one solution is to add padding to your XAxis

`padding={{ left: 50, right: 50 }}` - you can probably calculate how much you need based on your bar size

**Reply by softwareDev-git** [reply]
Great, thanks for the quick response.

**Reply by softwareDev-git** [reply]
@ckifer Apologies for continuing this discussion. Let me know if it is better to start a new discussion and I will create one.

I am attempting a solution where `Customized` component is used to render a `Rectangle` component instead of the `Bar` component. However, I am unsure on how to handle tooltip behaviour. Is it possible to use https://recharts.org/en-US/api/Rectangle#onMouseOver to show a tooltip? Any guidance on whether this is possible or not would be helpful.

See updated code sandbox that includes this solution - [Code sandbox for Recharts](https://codesandbox.io/p/sandbox/sandbox-for-recharts-chqn82?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxuc7nfe0006356ixtqlaplj%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxuc7nfe0002356ifef9yoys%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxuc7nfe0003356izcrp2hpp%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxuc7nfe0005356i8dwz04b4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxuc7nfe0002356ifef9yoys%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxuc7nfe0001356i6rpce7sf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxuk2c0n001n356ihets0d4o%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxuc7nfe0002356ifef9yoys%2522%252C%2522activeTabId%2522%253A%2522clxuk2c0n001n356ihets0d4o%2522%257D%252C%2522clxuc7nfe0005356i8dwz04b4%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxuc7nfe0004356ipr4aqib6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxuc7nfe0005356i8dwz04b4%2522%252C%2522activeTabId%2522%253A%2522clxuc7nfe0004356ipr4aqib6%2522%257D%252C%2522clxuc7nfe0003356izcrp2hpp%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxuc7nfe0003356izcrp2hpp%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

\
Below is the code snippet that replaces the `Bar` component.
```
{data.map((rectProps, index) => (
          <Customized
            key={index}
            component={(customisedChartProps) => {
              const { xAxisMap, yAxisMap } = customisedChartProps;
              const xScale = xAxisMap[0].scale;
              const yScale = yAxisMap[0].scale;
              const { timestamp, count, timestamp_end } = rectProps;
              return (
                <Rectangle
                  fill="lightgreen"
                  width={
                    xScale(timestamp_end - 0.25 * 3600000) - xScale(timestamp)
                  }
                  height={yScale(count) - yScale(0)}
                  x={xScale(timestamp)}
                  y={yScale(0)}
                />
              );
            }}
          />
        ))}
```

**Reply by ckifer** [reply]
Customized does not support Tooltip unfortunately - its usually used in conjunction with another component to render custom shapes. What I would do it continue to use Bar and instead provide a `shape` prop so you can customize the rectangle.

```
          shape={(props) => {
            return (
              <Rectangle
                fill="lightgreen"
                ...props
              />
            );
          }}
```

**Reply by softwareDev-git** [reply]
Understood. Thanks.
